crates/composefs/src/fsverity/mod.rs

```diff
@@ -72,10 +109,115 @@ pub fn compute_verity<H: FsVerityHashValue>(data: &[u8]) -> H {
-pub fn enable_verity<H: FsVerityHashValue>(fd: impl AsFd) -> Result<(), EnableVerityError> {
+pub fn enable_verity_raw<H: FsVerityHashValue>(fd: impl AsFd) -> Result<(), EnableVerityError> {
     ioctl::fs_ioc_enable_verity::<H>(fd)
 }
 
+/// Enable fs-verity on the given file, retrying if file is opened for writing.
+///
+/// This uses `enable_verity_raw()` and is subject to the same restrictions and features.
+///
+/// A common pattern with fsverity files is:
+///
+/// * Open a read-write file descriptor
+/// * Write data to the read-write file descriptor
+/// * Re-open the file descriptor as a new read-only descriptor
+/// * Close the read-write file descriptor
+/// * Enable fsverity on the read-only file descriptor
+///
+/// However, in a multi-threaded program, it is possible that another
+/// thread calls `fork()` while the read-write descriptor is valid,
+/// thus making a copy of the read-write descriptor.  If the forked
+/// process does not close the file descriptor either explicitly or by
+/// calling `exec()` via O_CLOEXEC, then attempting to enable fsverity
+/// on the read-only file descriptor will fail with ETXTBSY.  It is
+/// generally assumed that the file descriptor will be closed rather
+/// quickly under these circumstances, so this function will try to
+/// enable verity three times, pausing for one millisecond between
+/// attempts.
+pub fn enable_verity_with_retry<H: FsVerityHashValue>(
+    fd: impl AsFd,
+) -> Result<(), EnableVerityError> {
+    let mut attempt = 1;
+    loop {
+        match enable_verity_raw::<H>(&fd) {
+            Err(EnableVerityError::FileOpenedForWrite) if attempt < 3 => {
+                std::thread::sleep(std::time::Duration::from_millis(1));
+                attempt += 1;
+            }
+            other => return other,
+        }
+    }
+}
+
+/// Enable fs-verity on the given file.  If the given file cannot be
+/// enabled because it is opened as writable, then a new copy of the
+/// file will be returned instead.  No attempt is made to sync the
+/// copied file contents to disk, it is up to the caller to do so if
+/// desired.
+///
+/// Take special note that in the case where a copied file descriptor
+/// is returned, the returned file is created as a tempfile and is
+/// unlinked.  Presumably the caller should take care to make this
+/// file permanent, using a combination of `linkat` and `renameat` to
+/// replace the original file.
+///
+/// This uses `enable_verity_raw()` and `enable_verity_with_retry()`
+/// and is subject to the same restrictions.
+///
+/// # Arguments:
+/// * `dirfd`: A directory file descriptor, used to determine the placement (via O_TMPFILE) of the new file (if necessary).
+/// * `fd`: The file decriptor to enable verity on
+pub fn enable_verity_maybe_copy<H: FsVerityHashValue>(
+    dirfd: impl AsFd,
+    fd: OwnedFd,
+) -> Result<VerityFd, EnableVerityError> {
+    match enable_verity_with_retry::<H>(&fd) {
+        Ok(_) => Ok(VerityFd::Orig(fd)),
```

Comment:
I think a best practice is to spell this `Ok(())` - it looks weirder but ensures that we're only discarding a unit return value.